streaming/stream_transfer_task.cc

```diff
@@ -199,18 +199,25 @@ future<> send_mutation_fragments(lw_shared_ptr<send_info> si) {
         }();
 
         auto sink_op = [sink, si, got_error_from_peer] () mutable -> future<> {
-            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment>& sink) {
+            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, bool>& sink) {
                 return repeat([&sink, si, got_error_from_peer] () mutable {
                     return si->reader(db::no_timeout).then([&sink, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
                         if (mf && !(*got_error_from_peer)) {
                             frozen_mutation_fragment fmf = freeze(*s, *mf);
                             auto size = fmf.representation().size();
                             streaming::get_local_stream_manager().update_progress(si->plan_id, si->id.addr, streaming::progress_info::direction::OUT, size);
-                            return sink(fmf).then([] { return stop_iteration::no; });
+                            bool error = false;
+                            return sink(fmf, error).then([] { return stop_iteration::no; });
                         } else {
                             return make_ready_future<stop_iteration>(stop_iteration::yes);
                         }
                     });
+                }).handle_exception([&sink] (std::exception_ptr ep) mutable {
+                    // Notify the receiver the sender has failed
+                    bool error = true;
+                    return sink(frozen_mutation_fragment(bytes_ostream()), error).then([ep = std::move(ep)] () mutable {
```

Comment:
What if this `sink()` call fails, wouldn't we have the same problem?

Reply:
If sink() call fails, the close() call will fail as well (@gleb-cloudius can you confirm?), so the end of stream will not be generated.

Reply:
I have a better version to fix this issue. See the updates below.

Reply:
I don't think that has to be the case:
 1) `sink()` may fail on memory allocation, but `close()` may succeed.
 2) doesn't receiver close the stream when connection is dropped regardless of what state the sender is in? \cc @gleb-cloudius

Reply:
If sink() fails (for any reason) it suppose to close the underlying connection which will be seen as an error by a receiver.